A14-channelz.md

```diff
@@ -0,0 +1,1051 @@
+id must be distinct and may not be reused over the lifetime of the process for
+any entity.  This ensures that querying channelz will never result in referring
+to the wrong entity.  While ids are not allocated sequentially, they are
+ordered.  This enables pagination of the results when querying channelz.
+Since protobuf treats the number 0 as the default value, it is unused as an id.
+
+Along with the id, a human readable name can optionally be associated with each
+the id.  The id and the name together form a "reference".  Thus, Channels,
+Servers and Sockets are identified by their respective references.  These
+references are abbreviated ChannelRefs, SubchannelRefs, ServerRefs, and 
+SocketRefs.  Note that only the id is necessary to query channelz.
+
+The data representation of each ref:
+
+![message ref structure][message ref structure]
+
+
+## Channels and Subchannels
+
+Channels and Subchannels, or descendent channel, are hierarchically organized
+into a DAG structure.  The union of all channels and subchannels may not contain
+a cycle.  A descendent channel may have any number of descendent channels. Each
+descendent channel may also have any number of sockets.   However, a given 
+descendent channel cannot have heterogeneous children.  That is, a channel or
+subchannel may have descendent channels, or have sockets, but not both.
+
+A subchannel represents a channel that is load-balanced over.  When a channel
+has both subchannels and channels, the channels are not delegated to by the 
+superchannel.  An example would be a channel that is used by the ancestor 
+channel, but does not handle RPCs given by the application.
+
+Each channel has a ChannelRef reference which includes the channel id and an
+optional name.  The name is included for human consumption.  There are no
+restrictions on the name but it should be limited to a reasonable length.
+
+Each subchannel has a SubchannelRef reference which includes the subchannel id
+and an optional name.  The name is included for human consumption.  There are 
+no restrictions on the name but it should be limited to a reasonable length.
+
+![channel hierarchy 1][channel hierarchy 1]
+
+### Channel Data
+
+Channels and Subchannels include data about themselves, notably their initial
+parameters and their call stats.  Currently the data includes:
+
+* Channel state ( See: connectivity-semantics-and-api.md )
+* Channel trace
+* Channel target, if applicable.
+* Number of calls starts, succeeded, and failed.  Note this is slightly
+  different than the number of streams.  See the Socket Data section.
+* Time of the last call started.
+
+In general, each piece of data included is specific to the channel itself and
+NOT of its descendent channels.  This is to say that an ancestor channel
+is not an aggregator for descendent channels.  The target of a channel or
+subchannel may be different from that of its parent, or not present at all.  The
```

Comment:
Are there cases where the target of a subchannel will be different from that of its parent?  I think this is true for descendent channels (like an LB channel), but I don't think it is for subchannels.

Reply:
Probably, especially if two channels were created to different targets, but ended up with the same addresses.  The subchannels could be shared, but the original target would be relevant.  

In any case, some one may want to do it, even if implementations today don't.

Reply:
I think it's unlikely that we would ever have an architecture where a subchannel would be shared across channels with different targets, because the authority would be different.  And even if we did, we would probably want the subchannel to not show any target rather than showing one of the two targets randomly.  So I think we can support the ability for future implementations to do this by simply saying that the target of a subchannel may not be present.